pkg/cmd/util/output/backup_describer.go

```diff
@@ -284,6 +287,31 @@ func DescribeBackupStatus(ctx context.Context, kbClient kbclient.Client, d *Desc
 		d.Println()
 	}
 
+	if status.AsyncBackupItemOperationsAttempted > 0 {
+		if !details {
+			d.Printf("Async Backup Item Operations:\t%d of %d completed successfully, %d failed (specify --details for more information)\n", status.AsyncBackupItemOperationsCompleted, status.AsyncBackupItemOperationsAttempted, status.AsyncBackupItemOperationsFailed)
+			return
+		}
+
+		buf := new(bytes.Buffer)
+		if err := downloadrequest.Stream(ctx, kbClient, backup.Namespace, backup.Name, velerov1api.DownloadTargetKindBackupItemOperations, buf, downloadRequestTimeout, insecureSkipTLSVerify, caCertPath); err != nil {
```

Comment:
Actually, I think we *do* want to show this, since that's what will tell the user how much progress is made.

The backup repo will only update under two circumstances:
1) if the async controller moves a backup into a terminal state (Completed, PartiallyFailed) or registers an error response from a plugin (it won't persist every update to NCompleted)
2) If a user has explicitly asked for an update via `velero backup describe --details`

Interim changes that don't affect Backup completion or error count are only tracked in the controller unless a user creates a DownloadRequest. The DownloadRequest controller, when processing a new DownloadRequestfor item operation json will check for un-persisted changes and, if found, will trigger an upload before providing a download URL to the client.

Note that the Velero client does not open a connection to the Velero server directly. It creates a download request and then when that request is updated with a download URL, the velero client queries the BSL directly. There is no mechanism in the current Velero architecture for the velero client (which will often be running outside the cluster) to make a direct call into the Velero pod. Also note that the plugin operations DownloadRequest will only be generated if the user passes in `--details` -- any architecture change to provide a cluster-exposed API for direct queries from out-of-cluster velero clients is out of scope for this feature in 1.11.

The DownloadRequest interaction with the async controller to force upload in response to user request was a compromise made between, on the one hand, updating every async update (which, as you mention, would result in too many updates to the backup store), and on the other hand, not being able to provide users with current progress information when they explicitly ask for it.

Reply:
About the the processing of `force upload->download` in `velero backup describe --details` command, as the discussion in the community meeting and also a following local discussion, the current code is basically working, so we are ok to go with the current code. Meanwhile, we believe that there are better solutions practically.

Here I add my own analysis:
1. The current processing is working for the current Velero, but not in future. 
The progress info (i.e., total bytes, done bytes, etc.) are so important for a long-running activity, that it is a common requirement that the progress is retrieved very frequently. For example, vendors, who use Velero in the backend, may have a sophisticated UI by their own. In the UI, for a backup, they want to show a progress bar. The progress bar is displayed graphically from two factors -- the total bytes and done bytes, so the UI needs to query the progress from Velero every a couple of seconds.
For sure, we can quota the call to `velero backup describe --details` command with various methods, but how do we meet the requirement above?
2. The current restriction that it requires to write+read the backup storage in order to get the up-to-date progress is not due to a technical barrier, but our notion of the design -- the SnapshotBackup CR (and similar plugin specific objects) is not allowed to be visited by the async operation controller and Velero directly. 
For example, technically speaking, we can solve the problem by adding below logic in the client side:

      - We add some flexibility to detect the user of an async operation, let's call it userType
      - If the userType is datamover, we get the progress from the SnapshotBackup CR by querying the API server, something like `veleroClient.VeleroV1().SnapshotBackups.Get`
      - If the userType is xxx, we can use the user's specific way to get the progress similarly
      - If nothing is retrieved in the above try, we fallback to get it from the backup storage as the current code

This is totally feasible technically, and after some polish, I believe we can make it reasonable in logic without conflicting with the current notion of design

Reply:
It sounds like this particular subthread may be best discussed in a separate github issue, since implementation will be post-1.11 as it will require additional design work and API changes, but my initial thought is that to add the consumerType suggestions to the client code will require the velero client to be plugin-aware. Remember, SnapshotBackups mean nothing to the DownloadRequest controller or the velero client. Whatever solution we come up with also has to work with any other plugin, so we'll need to find a way for the plugin to be involved if we're going to pull this information from the APIServer locally. Also, this isn't about consumer type but plugin type. The same backup might have datamover async actions and custom async actions. For datamover actions, it would be SnapsohtBackups, but it might be a different resource for a different plugin. Also, since we have a defined interface for reporting progress on a single action from the plugin and on reporting progress for a group of actions to the client, this new work would need to fit into that framework. One possibility would be to implement streaming from the velero server directly. DownloadRequests for itemoperations could then be served a URL that streams from the velero pod instead of s3. The downside is that then we'd have to deal with exposing a public route to the velero pod and deal with certs/etc. This would not be a simple fix.

Another possibility would be for a new Velero CR and controller, analogous to the DownloadRequest controller that collects data directly for the user instead of relying on the object store. The downside here is that there are size limitations on kubernetes resources, so we may have to implement it with links to associated items -- maybe the resource links to a slice of configmaps. Something like this:
```
apiVersion: velero.io/v1
kind: AsyncOperationsRequest
metadata:
    name: foo
   ...
spec:
    backup: my-backup
```
Then the controller that watches this will create one or more configmaps with the equivalent content from itemoperations.json for async actions for this backup. The method of doing this would be as follows:
1) backups in terminal state: pull from object store
2) backups WaitingForPluginOperations: pull from the map managed by the async controller. No need to query the plugin directly, since the in-memory map has the latest that the controller has acted on and will be consistent with the Backup CR state.
3) backups in other states: we won't have operations yet for an InProgress or New backup, so it's just an empty list.

This will solve the problem for all async plugins without having to make a special case for datamover, and it will be an easy-ish add-on feature for 1.12, since it won't require changes to plugin or DownloadRequest API design -- it adds a completely new CRD instead.

Reply:
I agree that creating a new approach for the client to get whatever info from the Velero server (besides the current way to have the backup storage as an intermediation) will be increasingly required along with Velero's evolvement.
I also agree that this approach could be an enhancement that we can do in the future release, e.g., v1.12.

Furthermore, I suggest that we add below statements about the current limitation in the async progress monitor design, cc @reasonerjt :
`As the current design, if users want to get the up-to-date progress, Velero server needs to flush the data to backup storage and then Velero client download it on the client side. Therefore, this progress info is not recommended to be retrieved frequently, because this process is time consuming and may also be cost sensitive.`

Reply:
@Lyndon-Li I think we're in complete agreement here. I'd already talked to Shubham about the need to document the potential performance hit on `describe --details` -- lets pursue the alternate client interface with the velero pod as a design enhancement in the 1.12 time frame. I'll submit a design PR for that after we've wrapped up 1.11 implementation. I'm currently leaning towards the approach described above -- using a new AsyncOperationsRequest instead of a DownloadRequest which embeds the data  directly via configmaps.

Reply:
OK. Let's document this limitation in v1.11 and create a new design to fix this in 1.12.